src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -302,60 +346,138 @@ val qualityFormat: Option[FastqQualityFormat] = None,
+
+    if (this.outputFastqs) {
+      new FastqRecordWriter(prefix, pairedEnd)
+    }
+    else {
+      val readGroup = new SAMReadGroupRecord(sample.sampleId)
+      readGroup.setSample(sample.sampleName)
+      readGroup.setLibrary(sample.libraryId)
+      readGroup.setPlatform("Illumina")
+      sample.description.foreach(readGroup.setDescription)
+      platformUnit.foreach(readGroup.setPlatformUnit)
+      sequencingCenter.foreach(readGroup.setSequencingCenter)
+      predictedInsertSize.foreach(readGroup.setPredictedMedianInsertSize)
+      runDate.foreach(readGroup.setRunDate)
+      platformModel.foreach(readGroup.setPlatformModel)
+
+      val header: SAMFileHeader = new SAMFileHeader
+      header.addReadGroup(readGroup)
+      header.setSortOrder(sortOrder)
+      comments.foreach(header.addComment)
+
+      new SamRecordWriter(PathUtil.pathTo(prefix + ".bam"), header, this.umiTag)
+    }
+  }
+
+  /** Based on the type of quality scores coming in, converts them to a numeric byte[] in phred scale. */
+  private[fastq] def convertQuality(quals: Array[Byte], version: FastqQualityFormat, name: String) = {
+    version match {
+      case FastqQualityFormat.Standard => SAMUtils.fastqToPhred(quals)
+      case FastqQualityFormat.Solexa => solexaQualityConverter.convertSolexaQualityCharsToPhredBinary(quals)
+      case FastqQualityFormat.Illumina => solexaQualityConverter.convertSolexa_1_3_QualityCharsToPhredBinary(quals)
+    }
+    // Check that the converted qualities are same.
+    quals.foreach { qual =>
+      val uQual: Int = qual & 0xff
+      require(0 <= uQual && uQual <= SAMUtils.MAX_PHRED_SCORE, s"Base quality $uQual is not in the range 0 ... ${SAMUtils.MAX_PHRED_SCORE} for read: $name")
+    }
+  }
+}
+
+/** A writer than writes [[DemuxRecord]]s */
+private trait DemuxWriter extends Closeable {
+  def add(rec: DemuxRecord): Unit
+}
+
+/** A writer that writes [[DemuxRecord]]s as [[SAMRecord]]s. */
+private class SamRecordWriter(output: PathToBam,
+                              val header: SAMFileHeader,
+                              val umiTag: String) extends DemuxWriter {
+  private val writer = new SAMFileWriterFactory()
+    .setUseAsyncIo(DemuxFastqs.UseAsyncIo)
+    .setMaxRecordsInRam(DemuxFastqs.MaxRecordsInRamPerSamFileWriter)
+    .makeSAMOrBAMWriter(header, header.getSortOrder == SortOrder.unsorted, output.toFile)
+
+  private val rgId: String = this.header.getReadGroups.get(0).getId
+
+  def add(rec: DemuxRecord): Unit = {
+    val record  = new SAMRecord(header)
+    record.setReadName(rec.name)
+    record.setReadString(rec.bases)
+    record.setBaseQualities(rec.quals.getBytes)
+    record.setReadUnmappedFlag(true)
+    if (rec.pairedEnd) {
+      record.setReadPairedFlag(true)
+      record.setMateUnmappedFlag(true)
+      if (rec.readNumber == 1) record.setFirstOfPairFlag(true) else record.setSecondOfPairFlag(true)
+    }
+    record.setAttribute("BC", rec.sampleBarcode.orNull)
```

Comment:
For this and the molecularBarcode below why not use the usual?:
`rec.sampleBarcode.foreach(bc => record.setAttribute("BC", bc))`